picolibc.ld.in

```diff
@@ -162,9 +162,18 @@ SECTIONS
 	 * allocated at runtime. We're slightly abusing that
 	 * by placing the data in flash where it will be copied
 	 * into the allocate ram addresses by the existing
-	 * data initialization code in crt0
+	 * data initialization code in crt0.
+	 * BFD includes .tbss alignment when computing .tdata
+	 * alignment, but for ld.lld we have to explicitly pad
+	 * as it only guarantees usage as a TLS template works
+	 * rather than supporting this use-case.
 	 */
-	.tdata : @LDSCRIPT_ALIGN_WITH_INPUT@ {
+	.tdata : @LLD_START@ ALIGN(__tls_align) @LLD_END@ @BFD_START@ ALIGN_WITH_INPUT @BFD_END@ {
```

Comment:
I was hoping we could use the same expression for bot linkers, but bfd doesn't like `ALIGN(__tls_align)` or `ALIGN(MAX(ALIGNOF(.tdata), ALIGNOF(.tbss)))` here, as it complains the value is not a constant. I assume it just tries to assign those values earlier than lld, so they can't be resolved yet.

Reply:
It looks like there's a circular dependency -- you want `.tdata` to be aligned to `__tls_align`, but `__tls_align` depends upon `ALIGNOF(.tdata)`. I'm not sure how that could ever work.

Reply:
Maybe you just need to add `ALIGN_WITH_INPUT` support to lld?

Reply:
If I recall correctly, it's not `ALIGN_WITH_INPUT` that's causing the behaviour divergence here (lld has matching behaviour out of the box), but rather than ld.bfd implicitly includes the alignment of .tbss when computing that of .tdata.  I haven't looked at the linker script code in lld for the past few years but I believe this is not really a cyclic dependency since the alignment of all the input sections is known before final addresses are assigned, but I guess bfd evaluates the expression earlier.

Reply:
sure, but asking it to compute the `.tdata` alignment using a value that depends on the `.tdata` alignment doesn't seem like it will work.

Reply:
I just took a look at lld the code and I believe `LinkerScript::adjustOutputSections()` uses the alignment prior to assigning the final addresses, so this works as expected. @MaskRay is this valid to rely on this or am I relying on lld implementation details too much here and should find an alternative solution?